src/makefun/main.py

```diff
@@ -246,6 +246,8 @@ def create_function(func_signature,             # type: Union[str, Signature]
     if isinstance(func_signature, str):
         # transform the string into a Signature and make sure the string contains ":"
         func_name_from_str, func_signature, func_signature_str = get_signature_from_string(func_signature, evaldict)
+        if '__signature__' in attrs:
```

Comment:
```suggestion
        # If the signature is specified as a string, `__signature__` will not be correct in this case, since it
        # will be a string. If it was computed in some other fashion (i.e. is an instance of inspect.Signature),
        # then there is no need to update it.
        if '__signature__' in attrs:
```

> Is this line tested in your test_issues.py by the way ?

This was added because of a test failure in a test which specified the signature as a string. Commenting out these lines causes a failure in `tests/test_advanced.py::tests_wraps_sigchange`, with error `TypeError: unexpected object '(a, b=0)' in __signature__ attribute`. 

I can add to the test in test_issues.py to also reproduce the error there, but it would be duplicative of `tests_wraps_sigchange`. I'm not sure what the right way to handle annotations and forward references in a string signature is, though setting `__wrapped__` / `__signature__` seems good for consistency.

Reply:
Apologies @lucaswiman , I have some trouble understanding the current code comment: how does the fact that `func_signature` is a string, relate to the fact that `__signature__` is in attrs ?

From PEP362:

>  the user can manually cache a Signature by storing it in the __signature__ attribute.
> It is better to reserve the `__signature__` attribute for the cases when there is a need to explicitly set to a `Signature` object that is different from the actual one

So we should probably not set `attrs['__signature__']` in this function at all, only in `wraps` (if the bug with forward references is only happening in `wraps`). Alternately, if the bug is also happening with `create_function`, it is mandatory to clearly explain in the docstring of `create_function` in which conditions it will be set, and why.

Let's discuss this directly in the ticket before coming back to the PR: https://github.com/smarie/python-makefun/issues/85#issuecomment-1176318426

Thanks!

Reply:
The problem is that if you pass a string signature to wraps, then `__signature__` ends up as a string as well. `wraps` knows whether it needs to set `__signature__`, so I don't think you can move that logic here. _But_, you don't have access to the compiled signature object in `wraps`, only in this method. 

It seems like the code is just a bit awkwardly factored to handle this edge case, and this seemed like the least disruptive place to put this logic.

So I see two options:
1. Since it is _always_ an error to set `__signature__` to a string value, we could set `attrs['__signature__'] = get_signature_from_string(attrs['__signature__'], evaldict)[1]` when `attrs[__signature__]` is a string. 
2. Delay setting both `__wrapped__` and `__signature__` until after the call to this function in wraps. 

I guess (2) is more in line with what you're asking for. (1) seems a bit simpler, and I think it would never do the wrong thing. (I'm sort of guessing what the right behavior is here, since I don't have a good understanding of the use cases for specifying the signature as a string.)

Reply:
@smarie I think maybe what was confusing is that `attrs['__signature__']` will happen to be the same as `func_signature` when it is specified, but I suppose there's no logical reason to assume that would _always_ be the case. (Though in fact it would be a bug if they differed when both are defined.) It's also an error if either of them are strings, so they both need to be converted.

In 0fe6a7e067c74b791b545f36bc6224f528c92344, I updated the code to do the simplest obviously-correct thing of treating string `__signature__` entries as completely separate from `func_signature` and separately evaluating them as a string. I think this is less efficient, but in the interest of moving this PR along, I updated the logic. Could you please take a look?